draft-wang-ppm-dap-taskprov.md

```diff
@@ -56,76 +60,330 @@ collector). Thus it is up to each deployment of DAP to securely implement its
+Not Found and an error of type `unrecognizedTask`. The response MAY include a
+Retry-After header field to suggest a pulling interval to the collector.
+
+> Alternatively the leader can just response with 303 and continue waiting.
 
 # Helper Behavior
 
-> TODO(wangshan) Say what the Helper does when it receives a report share with
-> this extension. (Same questions as above.)
+Helper should have received out-of-band from leader any information that is not
+always tied to a particular task, for e.g. `vdaf_verify_key`. Similar to
+leader, helper may not know the task ID of the current task before receiving
+the first `AggregateInitializeReq` message from the leader.
+
+## Change to helper aggregate sub-protocol
 
+Upon receipt of a `AggregateInitializeReq`, helper reads the extension
+codepoint in `extension_type`. If helper does not support this extension, it
+SHOULD abort the aggregate protocol and alert the leader with error
+"unrecognizedMessage".
+
+If helper supports `task-prov` extension, it should first check if the task ID
+already exists, if so helper continues to the rest of helper initialization as
+usual (see helper initialization in {{?DAP=I-D.draft-ietf-ppm-dap-01}}.)
+Note if the existing task's configuration is different from the one in report
+share extension, HPKE decryption will fail due to mismatched AAD.
+
+If the task ID has not been seen before, helper should read and decode report
+share's `extension_data` with the `TaskConfig` schema. If the decode failed,
+it MUST abort the aggregate protocol and alert the leader with error
+"unrecognizedMessage".
+
+If the decode succeeds, helper should create a new task using the task ID from
+the decoded report share, and save task configuration with the newly created
+task. In particular, helper should pass `vdaf_config` to the VDAF initializer,
+based on `vdaf_type` in `TaskConfig`. At this point, the task provision step has
+completed, and helper should continue to the rest of helper initialization.
 
 # Collector Behavior
 
 > TODO(wangshan) Describe if/how this extension impacts Collector behavior.
 
+If Collector supports `task-prov` extension and receives a HTTP status code
+404 Not Found with error type `unrecognizedTask` after sending a `CollectReq`,
+it SHOULD retry with the same `CollectReq`, potentially using an interval from
+the Retry-After header in the received response.
+
+# Implementation and Operational Considerations
+
+> NOTE(shan) Do we want to include this section in the beginning?
+
+A "task" now means the same task ID and same task configuration, if a malicious
+client changes the task ID or task configuration, its report will be aggregated
+in a different task, with other poison reports from the same malicious attack.
+The "good" reports will not be polluted.
+
+The in-band task provision mechanism is easy to implement with streaming
+framework that has `groupBy` operator. In fact, task as an object doesn't have
```

Comment:
> streaming framework that has `groupBy` operator.

This is too vague. If you want to suggest a particular implementation strategy, then you'll need to elaborate more here.

Reply:
I don't think this paragraph is essential, let me remove this from the first draft and add details later